modules/volume/pulseaudio/pulse.go

```diff
@@ -115,131 +70,142 @@ func Source(sourceName string) volume.Provider {
+type sinkController struct {
+	client     *proto.Client
+	deviceName string
 }
 
-func openFallbackDevice(conn *dbus.Conn, core dbus.BusObject, deviceType deviceType) (dbus.BusObject, error) {
-	path, err := core.GetProperty("org.PulseAudio.Core1.Fallback" + deviceType.String())
-	if err != nil {
-		return nil, err
+type sourceController struct {
+	client     *proto.Client
+	deviceName string
+}
+
+func (c *sinkController) SetVolume(newVol int64) (err error) {
+	return c.client.Request(&proto.SetSinkVolume{
+		SinkIndex:      proto.Undefined,
+		SinkName:       c.deviceName,
+		ChannelVolumes: proto.ChannelVolumes{uint32(newVol)},
+	}, nil)
+}
+
+func (c *sourceController) SetVolume(newVol int64) (err error) {
+	return c.client.Request(&proto.SetSourceVolume{
+		SourceIndex:    proto.Undefined,
+		SourceName:     c.deviceName,
+		ChannelVolumes: proto.ChannelVolumes{uint32(newVol)},
+	}, nil)
+}
+
+func (c *sinkController) SetMuted(muted bool) (err error) {
+	return c.client.Request(&proto.SetSinkMute{
+		SinkIndex: proto.Undefined,
+		SinkName:  c.deviceName,
+		Mute:      muted,
+	}, nil)
+}
+
+func (c *sourceController) SetMuted(muted bool) (err error) {
+	return c.client.Request(&proto.SetSourceMute{
+		SourceIndex: proto.Undefined,
+		SourceName:  c.deviceName,
+		Mute:        muted,
+	}, nil)
+}
+
+func getVolume(client *proto.Client, deviceName string, deviceType deviceType) (vol volume.Volume, err error) {
+	switch deviceType {
+	case SinkDevice:
+		return getVolumeSink(client, deviceName)
+	case SourceDevice:
+		return getVolumeSource(client, deviceName)
+	default:
+		panic(fmt.Sprintf("unexpected device type %v", int(deviceType)))
 	}
-	return openDevice(conn, core, path.Value().(dbus.ObjectPath), deviceType)
 }
 
-func getVolume(device dbus.BusObject) (volume.Volume, error) {
-	max, err := device.GetProperty("org.PulseAudio.Core1.Device.BaseVolume")
+func getVolumeSink(client *proto.Client, deviceName string) (vol volume.Volume, err error) {
+	repl := proto.GetSinkInfoReply{}
+	err = client.Request(&proto.GetSinkInfo{SinkIndex: proto.Undefined, SinkName: deviceName}, &repl)
 	if err != nil {
-		return volume.Volume{}, err
+		return
 	}
-	maxVol := int64(max.Value().(uint32))
+	return makeVolume(repl.ChannelVolumes, repl.BaseVolume, repl.Mute, &sinkController{client, deviceName}), nil
+}
 
-	vol, err := device.GetProperty("org.PulseAudio.Core1.Device.Volume")
+func getVolumeSource(client *proto.Client, deviceName string) (vol volume.Volume, err error) {
+	repl := proto.GetSourceInfoReply{}
+	err = client.Request(&proto.GetSourceInfo{SourceIndex: proto.Undefined, SourceName: deviceName}, &repl)
 	if err != nil {
-		return volume.Volume{}, err
+		return
 	}
+	return makeVolume(repl.ChannelVolumes, repl.BaseVolume, repl.Mute, &sourceController{client, deviceName}), nil
+}
 
+func makeVolume(channelVolumes proto.ChannelVolumes, baseVolume proto.Volume, mute bool, controller volume.Controller) volume.Volume {
+	maxVol := int64(baseVolume)
```

Comment:
Okay, I've just noticed that this part is wrong. 100% volume should be `PA_VOLUME_NORM` (0x10000), not base volume.

Reply:
Fixed